### PR TITLE
Fixes Char Prefs Not Saving/Spawning Correctly

### LINKED ISSUE #8386: Lizards can't use digitigrade legs
## Round ID & Round Date:

Don't remember the exact round IDs. (26.01.2023)

Tested it locally too

## Reproduction:
Pick a lizard and select digi legs:
Untitled

Doesn't work when round starts:
Untitled1

Ashwalkers still use digitigrade legs.

### LINKED ISSUE #8399: apid wings are invisible
Round ID: 42448 1-28-23

## Testmerges:merged:
[#8387: 'Derelict shuttle ruins (Custom shuttles save and appear as ruins)' by PowerfulBacon at commit de67265d4b](https://www.github.com/beestation/beestation-hornet/pull/8387)

## Reproduction:
play apid, look at your back, see lack of wings, see that you can still dash, observe how *wing does not work

### PR DESCRIPTION
**Pull request by DatBoiTim**
## About The Pull Request
Fixes #8386 

## Why It's Good For The Game
FEEEEEX

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Local test Digit

Local test Digit Spawn

Confirmed fixes https://github.com/BeeStation/BeeStation-Hornet/issues/8386

</details>

## Changelog
:cl: DatBoiTim, Rkz
fix: fixed a few issues related to preferences not saving/spawning you in right
/:cl:

### CODE REVIEW

**Comment by itsmeow** [comment]
where my testing evidence fool

**Comment by DatBoiTim** [comment]
> where my testing evidence fool

Local test Digit

Local test Digit Spawn

Confirmed fixes #8386

**Comment by DatBoiTim** [comment]
Testing Fix for Bee wings

**Comment by DatBoiTim** [comment]
Bee wings back
Bee wings working but still invisible.

**Comment by DatBoiTim** [comment]
Bee wings sprite is gone

**Comment by DatBoiTim** [comment]
#2143

This broke the wing sprites.

**Comment by DatBoiTim** [comment]
Beewing issues might be something else

**Review by itsmeow — APPROVED** [review]

File: code/modules/client/preferences2/character_save.dm [hunk 1]
```diff
@@ -469,6 +469,7 @@
 		pref_species = new /datum/species/human
 		save(usr.client, async = FALSE) // This entire proc is called a lot at roundstart, and we dont want to lag that
 
+	character.dna.features = features.Copy()
 	character.set_species(chosen_species, icon_update = FALSE, pref_load = TRUE)
 	character.dna.features = features.Copy()
```
Comment (Crossedfall) [inline code/modules/client/preferences2/character_save.dm:474]: This cannot be the best way to handle this. What _actually_ causes all this stuff to break?

Comment (Rukofamicom) [reply code/modules/client/preferences2/character_save.dm:474]: > What actually causes all this stuff to break?

image

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
This solution feels so jank, I'm not merging this without an explanation of why this works and why there is no proper way to solve this.

**Comment by DatBoiTim** [comment]
> This solution feels so jank, I'm not merging this without an explanation of why this works and why there is no proper way to solve this.

Ok, so few things

# RCA

set_species replaces the bodyparts according to the dna.features of a passed carbon.

The use of digitigrade legs, etc. are added during the species change according to the features currently stored on the carbon, and not the features as read straight from the preferences.

Thus the current execution goes on master goes like this:

1) Make a blank body, with default features*
2) set_species of body, all body parts are determined by both species and the features on the body**
3) set the features of the body to have digi legs

*Digi Legs are non-default value
**if features do not indicate use of digi legs, normal legs are added 

# Analysis of master/character_save.dm line 473 

I have no idea why the inclusion of `character.dna.features = features.Copy()` at line 473 of character_save.dm fixes spines showing in preview.

I've put a breakpoint for every instance of spines related to lizards, counted out the number of skips until the change happens, run another test, and started drilling down on the call stack right before it happens.

This change happens somewhere within /client/proc/send_resources(), most notably, around line 964. I do not know why removing the `character.dna.features = features.Copy()` at line 473 causes the change not to happen.

File: code/modules/client/preferences2/character_save.dm (hunk 1 above)
Comment (DatBoiTim) [reply code/modules/client/preferences2/character_save.dm:474]: Long story short, without 
`character.dna.features = features.Copy()` before
`character.set_species(chosen_species, icon_update = FALSE, pref_load = TRUE)`
When `character.set_species(chosen_species, icon_update = FALSE, pref_load = TRUE)` will not have features updated to spawn the person in with digi legs.

Now, as to why `character.dna.features = features.Copy()` after
`character.set_species(chosen_species, icon_update = FALSE, pref_load = TRUE)` fixes the preview screen, I do not know. The update occurs somewhere within /client/proc/send_resources(), most notably, around line 964. I cannot drill down any further than this.
 I do not know why this update doesn't happen without `character.dna.features = features.Copy()` after `character.set_species(chosen_species, icon_update = FALSE, pref_load = TRUE)`.

**Comment by itsmeow** [comment]
Superseded by #8505

**Pull request closed without merging**